yt/data_objects/static_output.py

```diff
@@ -701,42 +710,83 @@ def setup_deprecated_fields(self):
+                f"Dataset object {self} has a tuple for its geometry attribute. "
+                "This is interpreted as meaning the first element is the actual geometry string, "
+                "and the second represents an arbitrary axis order. "
+                "This will stop working in a future version of yt.\n"
+                "If you're loading data using yt.load_* functions, "
+                "you should be able to clear this warning by using the axis_order keyword argument.\n"
+                "Otherwise, if your code relies on this behaviour, please reach out and open an issue:\n"
+                "https://github.com/yt-project/yt/issues/new",
+                since="4.2",
+                stacklevel=2,
+            )
+            self.geometry, axis_order = self.geometry
+        if type(self.geometry) is str:
+            issue_deprecation_warning(
+                f"Dataset object {self} has a raw string for its geometry attribute. "
+                "In yt>=4.2, a yt.geometry.geometry_enum.Geometry member is expected instead. "
+                "This will stop working in a future version of yt.\n",
+                since="4.2",
+                stacklevel=2,
+            )
+            self.geometry = Geometry.from_str(self.geometry)
         if isinstance(self.geometry, CoordinateHandler):
-            # I kind of dislike this.  The geometry field should always be a
-            # string, but the way we're set up with subclassing, we can't
-            # mandate that quite the way I'd like.
-            self.coordinates = self.geometry
-            return
-        elif callable(self.geometry):
-            cls = self.geometry
-        elif self.geometry == "cartesian":
+            issue_deprecation_warning(
+                f"Dataset object {self} has a CoordinateHandler object for its geometry attribute. "
+                "In yt>=4.2, a yt.geometry.geometry_enum.Geometry member is expected instead. "
+                "This will stop working in a future version of yt.\n",
+                since="4.2",
+                stacklevel=2,
+            )
+            _class_name = type(self.geometry).__name__
+            if not _class_name.endswith("CoordinateHandler"):
+                raise RuntimeError(
+                    "Expected CoordinateHandler child class name to end with CoordinateHandler"
+                )
+            _geom_str = _class_name[: -len("CoordinateHandler")]
+            self.geometry = Geometry.from_str(_geom_str)
+            del _class_name, _geom_str
+
+        # end compatibility layer
+        if not isinstance(self.geometry, Geometry):
+            raise TypeError(
+                "Excpected dataset.geometry attribute to be of "
+                "type yt.geometry.geometry_enum.Geometry\n"
+                f"Got {self.geometry=} with type {type(self.geometry)}"
+            )
+
+        cls: Type[CoordinateHandler]
+        if self.geometry is Geometry.CARTESIAN:
```

Comment:
a dict doesn't work as a replacement for an enum because it's mutable (well, I guess a enum is also technically mutable, but it's clearly not meant to be mutated). The combination of `StrEnum` + `assert_never` can be checked with mypy, no other formalism allows this decisive feature, that I know of.

Reply:
What I meant more was, mapping the enum as the key in a dict to the class as the value.

Reply:
Ah, well here this would indeed work if all we were doing was selecting a class, unfortunately there's also the matter of setting `self.no_cgs_equiv_length`